crates/sui-display/src/v2/lexer.rs

```diff
@@ -0,0 +1,846 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+#![allow(dead_code)]
+
+use std::fmt;
+
+/// Lexer for Display V2 format strings. Format strings are a mix of text and expressions.
+/// Expressions are enclosed in braces and may contain multiple alternates, separated by pipes, and
+/// each containing nested field, vector, or dynamic field accesses.
+#[derive(Debug)]
+pub(crate) struct Lexer<'s> {
```

Comment:
maybe not worth it at this point since you've already written this, but i've found i've really liked using `winnow` for doing this. I've felt it makes things much more readable about the structure of what is supported.